File: swift/Sources/AetherNetProtocol/Identity/EridDirectory.swift
// SPDX-License-Identifier: MIT

import Foundation

/// Resolves rotating ``EphemeralRoutingId`` (ERID) wire addresses to and from the stable
/// peer identities behind them — the piece that lets an ESTABLISHED relationship follow a
/// peer's rotating address while an outsider cannot.
///
/// A node derives its OWN secret routingKey once (via
/// ``EphemeralRoutingId/deriveRoutingKey(_:)``) and shares it with a peer INSIDE the
/// established Signal session — never on the wire. Each side stores the other's routingKey
/// here, so either can compute the other's current ERID for addressing and reverse-resolve
/// an inbound ERID back to the peer it belongs to. An outsider holds no routingKey and can
/// do neither. Port of the C# reference (`src/AetherNet.Core/Identity/EridDirectory.cs`).
public final class EridDirectory {

    private let myRoutingKey: [UInt8]
    private let epochSeconds: Int64
    private let eridLength: Int
    private var peerKeys: [String: [UInt8]] = [:]

    /// - Parameter myRoutingKey: this node's secret routingKey (from
    ///   ``EphemeralRoutingId/deriveRoutingKey(_:)``).
    /// - Throws: ``EphemeralRoutingId/EphemeralRoutingIdError/emptyRoutingKey`` if
    ///   `myRoutingKey` is empty, or ``.invalidEpochSeconds`` if `epochSeconds <= 0`.
    public init(
        _ myRoutingKey: [UInt8],
        epochSeconds: Int64 = EphemeralRoutingId.defaultEpochSeconds,
        eridLength: Int = EphemeralRoutingId.defaultLength
    ) throws {
        guard !myRoutingKey.isEmpty else {
            throw EphemeralRoutingId.EphemeralRoutingIdError.emptyRoutingKey
        }
        guard epochSeconds > 0 else {
            throw EphemeralRoutingId.EphemeralRoutingIdError.invalidEpochSeconds
        }
        self.myRoutingKey = myRoutingKey
        self.epochSeconds = epochSeconds
        self.eridLength = eridLength
    }

    /// Our own current ERID for the epoch containing `unixSeconds` — the address we present
    /// on the wire this window.
    public func myErid(_ unixSeconds: Int64) throws -> String {
        try EphemeralRoutingId.derive(
            myRoutingKey, unixSeconds: unixSeconds,
            epochSeconds: epochSeconds, length: eridLength
        )
    }

    /// Store a peer's routingKey, learned inside an established session. Idempotent; a
    /// later call replaces an earlier key for the same peer.
    /// - Precondition: `peerUhid` and `peerRoutingKey` are non-empty.
    public func rememberPeer(_ peerUhid: String, routingKey peerRoutingKey: [UInt8]) {
        precondition(!peerUhid.isEmpty, "peerUhid cannot be empty")
        precondition(!peerRoutingKey.isEmpty, "peerRoutingKey cannot be empty")
        peerKeys[peerUhid] = peerRoutingKey
    }

    /// Forget a peer (session torn down / excommunicated). Returns false if unknown.
    @discardableResult
    public func forgetPeer(_ peerUhid: String) -> Bool {
        peerKeys.removeValue(forKey: peerUhid) != nil
    }

    /// The current ERID a known peer presents this epoch, or `nil` if we hold no key for
    /// them.
    public func eridForPeer(_ peerUhid: String, unixSeconds: Int64) throws -> String? {
        guard let key = peerKeys[peerUhid] else { return nil }
        return try EphemeralRoutingId.derive(
            key, unixSeconds: unixSeconds,
            epochSeconds: epochSeconds, length: eridLength
        )
    }

    /// Reverse-resolve an inbound wire ERID to the stable peer UHID behind it for the given
    /// epoch, or `nil` if no known peer currently presents it. O(n) over known peers — a
    /// node's actual relationship count.
    public func resolvePeer(_ erid: String, unixSeconds: Int64) throws -> String? {
        guard !erid.isEmpty else { return nil }
        for (uhid, key) in peerKeys {
            let candidate = try EphemeralRoutingId.derive(
                key, unixSeconds: unixSeconds,
                epochSeconds: epochSeconds, length: eridLength
            )
            if candidate == erid { return uhid }
        }
        return nil
    }

    /// Number of peers whose routingKey we currently hold.
    public var knownPeerCount: Int { peerKeys.count }
}

File: src/AetherNet.Core/Identity/EridDirectory.cs
// SPDX-License-Identifier: MIT

using System.Collections.Concurrent;

namespace AetherNet.Identity;

/// <summary>
/// Resolves rotating <see cref="EphemeralRoutingId"/> (ERID) wire addresses to and from the stable
/// peer identities behind them — the piece that lets an ESTABLISHED relationship follow a peer's
/// rotating address while an outsider cannot.
///
/// <para>A node derives its OWN secret <c>routingKey</c> once (from its identity secret, via
/// <see cref="EphemeralRoutingId.DeriveRoutingKey"/>) and shares it with a peer <em>inside</em> the
/// established Signal session — never on the wire. Each side stores the other's routingKey here, so
/// either can compute the other's current ERID for addressing, and reverse-resolve an inbound ERID
/// back to the peer it belongs to. An outsider holds no routingKey and can do neither: to a passive
/// observer the ERID is an opaque value that rotates every epoch with no cross-window linkage.</para>
///
/// <para>Sharing the routingKey (rather than a single ERID) is deliberate: a party you have an
/// established relationship with is, by definition, allowed to recognise you over time. The ERID
/// hides you from <em>outsiders</em>, not from peers you have chosen to talk to. (A future hardening
/// can share only a bounded window of upcoming ERIDs instead, trading convenience for forward
/// unlinkability against a later-compromised peer.)</para>
///
/// <para>This is the in-memory directory only — additive, off-wire, and behind the eventual
/// <c>erid/v1</c> capability. It changes no serialized bytes on its own.</para>
/// </summary>
public sealed class EridDirectory
{
    private readonly byte[] _myRoutingKey;
    private readonly int _epochSeconds;
    private readonly int _eridLength;

    // peerUhid -> that peer's secret routingKey, learned inside an established session.
    private readonly ConcurrentDictionary<string, byte[]> _peerKeys = new(StringComparer.Ordinal);

    /// <param name="myRoutingKey">This node's secret routingKey — derive it with
    /// <see cref="EphemeralRoutingId.DeriveRoutingKey"/> from the identity secret. Copied defensively.</param>
    public EridDirectory(
        byte[] myRoutingKey,
        int epochSeconds = EphemeralRoutingId.DefaultEpochSeconds,
        int eridLength = EphemeralRoutingId.DefaultLength)
    {
        ArgumentNullException.ThrowIfNull(myRoutingKey);
        if (myRoutingKey.Length == 0)
            throw new ArgumentException("myRoutingKey cannot be empty.", nameof(myRoutingKey));
        if (epochSeconds <= 0)
            throw new ArgumentOutOfRangeException(nameof(epochSeconds));

        _myRoutingKey = (byte[])myRoutingKey.Clone();
        _epochSeconds = epochSeconds;
        _eridLength = eridLength;
    }

    /// <summary>Our own current ERID for the epoch containing <paramref name="unixSeconds"/> —
    /// the address we present on the wire this window.</summary>
    public string MyErid(long unixSeconds)
        => EphemeralRoutingId.Derive(_myRoutingKey, unixSeconds, _epochSeconds, _eridLength);

    /// <summary>Store a peer's routingKey, learned inside an established session. Idempotent;
    /// a later call replaces an earlier key for the same peer (e.g. after a re-key).</summary>
    public void RememberPeer(string peerUhid, byte[] peerRoutingKey)
    {
        ArgumentException.ThrowIfNullOrEmpty(peerUhid);
        ArgumentNullException.ThrowIfNull(peerRoutingKey);
        if (peerRoutingKey.Length == 0)
            throw new ArgumentException("peerRoutingKey cannot be empty.", nameof(peerRoutingKey));
        _peerKeys[peerUhid] = (byte[])peerRoutingKey.Clone();
    }

    /// <summary>Forget a peer (session torn down, or peer excommunicated). Returns false if unknown.</summary>
    public bool ForgetPeer(string peerUhid) => _peerKeys.TryRemove(peerUhid, out _);

    /// <summary>The current ERID a known peer presents this epoch, or null if we hold no key for them.</summary>
    public string? EridForPeer(string peerUhid, long unixSeconds)
        => _peerKeys.TryGetValue(peerUhid, out var key)
            ? EphemeralRoutingId.Derive(key, unixSeconds, _epochSeconds, _eridLength)
            : null;

    /// <summary>
    /// Reverse-resolve an inbound wire ERID to the stable peer UHID behind it for the given epoch,
    /// or null if no known peer currently presents it. O(n) over known peers — a node's actual
    /// relationship count (tens–hundreds); a caller on a hot path can cache the ERID→peer map per
    /// epoch (it only changes at a 15-minute boundary).
    /// </summary>
    public string? ResolvePeer(string erid, long unixSeconds)
    {
        if (string.IsNullOrEmpty(erid)) return null;
        foreach (var kvp in _peerKeys)
        {
            var candidate = EphemeralRoutingId.Derive(kvp.Value, unixSeconds, _epochSeconds, _eridLength);
            if (string.Equals(candidate, erid, StringComparison.Ordinal))
                return kvp.Key;
        }
        return null;
    }

    /// <summary>Number of peers whose routingKey we currently hold.</summary>
    public int KnownPeerCount => _peerKeys.Count;
}
